### PR TITLE
Implemented LivingKnockBackEvent event and hook

### PR DESCRIPTION
**Pull request by MercurialPony**
## **Changes**
Added _LivingKnockBackEvent_ event, added _onLivingKnockBack_ method to _ForgeHooks_, added _onLivingKnockBack_ check and values into _EntityLivingBase_ _knockBack_ method. Added _KnockBackHookTest_ test mod.

## **Why?**
This event fixes the problems with hard-coded vanilla knock back. Modders can now cancel, change or add their own knock back to entities with quite a few parameters at their disposal. I've seen a few threads asking for this type of event before.
This addition is absolutely necessary to me as currently I'm creating a mod that adds a few new attributes for items including a knock back attribute. Using events like _LivingAttackEvent_ to cancel knock back is very inefficient, sometimes inconsistent and potentially unstable.

## **How does it work?**
It's as simple as any other event in forge. The _LivingKnockBackEvent_ contains a few handy parameters like the attacker, strength and ratios which can be changed resulting in a highly flexible event.
I added a test mod _KnockBackHookTest_ for anyone doubting the hook's functionality.

### CODE REVIEW

File: patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch
Comment (bs2609) [inline patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch]: Don't change imports in patches.

File: patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch
Comment (bs2609) [inline patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch]: The formatting here doesn't match, check spaces are being used.

File: patches/minecraft/net/minecraft/item/Item.java.patch
Comment (bs2609) [inline patches/minecraft/net/minecraft/item/Item.java.patch]: You aren't really changing the contents of this file, so leave it out of the PR.

File: src/main/java/net/minecraftforge/common/ForgeHooks.java
Comment (bs2609) [inline src/main/java/net/minecraftforge/common/ForgeHooks.java]: As you are negating the result when using this function, it'd be better to avoid the double-negation by just returning the result here directly.

File: src/main/java/net/minecraftforge/event/entity/living/LivingKnockBackEvent.java
Comment (bs2609) [inline src/main/java/net/minecraftforge/event/entity/living/LivingKnockBackEvent.java]: Should be `LivingKnockBackEvent`?

File: src/main/java/net/minecraftforge/event/entity/living/LivingKnockBackEvent.java
Comment (bs2609) [inline src/main/java/net/minecraftforge/event/entity/living/LivingKnockBackEvent.java]: These values can be `final` as well.

File: src/test/java/net/minecraftforge/debug/KnockBackHookTest.java
Comment (bs2609) [inline src/test/java/net/minecraftforge/debug/KnockBackHookTest.java]: Add a `version = "1.0"` element here, to prevent a warning being logged.

File: src/test/java/net/minecraftforge/debug/KnockBackHookTest.java
Comment (bs2609) [inline src/test/java/net/minecraftforge/debug/KnockBackHookTest.java]: Can you make this mod be disabled by default?

File: src/main/java/net/minecraftforge/event/entity/living/LivingKnockBackEvent.java
Comment (bs2609) [inline src/main/java/net/minecraftforge/event/entity/living/LivingKnockBackEvent.java:20]: This file needs a license header.

**Review by bs2609 — COMMENTED** [review]
The hook looks ok, but there are a few style points.

Make sure to use spaces for indentation for consistency with the rest of the codebase.

Also try to keep changes in patches strictly to what you need to implement your hook: don't touch imports and only commit the patch files you actually worked on.

**Comment by Melonslise** [comment]
Thanks for the feedback, I'll get to fixing the issues right away. I made some big changes will allow the event's values to be changed. Could you re-review the _LivingKnockBackEvent_, _EntityLivingBase_ and _ForgeHooks_ classes?

**Comment by Melonslise** [comment]
@bs2609 Thanks for the feedback, I'll get to fixing the issues right away. However, I made some big changes that allow the event's values to be changed. Could you re-review the LivingKnockBackEvent, EntityLivingBase and ForgeHooks classes?

**Comment by bs2609** [comment]
I think a good way to handle allowing changes to the values is to patch in a block of code like this at the start of the `knockBack` method:
```java
net.minecraftforge.event.entity.living.LivingKnockBackEvent event
    = net.minecraftforge.common.ForgeHooks.onLivingKnockBack(this, entityIn, strength, xRatio, zRatio);
if (event.isCanceled()) return;
strength = event.getStrength(); xRatio = event.getRatioX(); zRatio = event.getRatioZ();
```

Have the hook method construct and post the event as before, but then return the event itself, rather than a boolean. Then you can overwrite the passed in parameters with the new values from the event.

The reason I'd suggest doing it this way is to keep all the code changes in one block, and to avoid needing to touch the vanilla logic.

**Comment by MercurialPony** [comment]
@bs2609 I fixed all the issues you pointed out. Will you approve now? When will these additions be merged with forge?

File: patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch
Comment (Draco18s) [inline patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch]: You have unused imports

File: patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch
Comment (Draco18s) [inline patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch]: You should use a fully qualified name, rather than an import here; `net.minecraftforge.event.entity.living.LivingKnockBackEvent event = ...`

File: src/main/java/net/minecraftforge/common/ForgeHooks.java
Comment (Draco18s) [inline src/main/java/net/minecraftforge/common/ForgeHooks.java]: You've reorganized the imports here, these lines have shifted place. Don't do this, it makes the patch files larger for no reason.

File: patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch
Comment (Draco18s) [inline patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch]: You have added a space here `(double) p_70653_2_` which causes additional lines in the patch file that do precisely nothing.

**Review by Draco18s — CHANGES_REQUESTED** [review]
Some more cleanup required before it'll be approved. Lex is very strict about the standards: the smaller a patch file is, the better.

File: src/main/java/net/minecraftforge/common/ForgeHooks.java
Comment (covers1624) [reply src/main/java/net/minecraftforge/common/ForgeHooks.java]: This is a forge class. Imports should be added but not organized, its entirely up to the users IDE as to how they are ordered.

File: patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch
Comment (covers1624) [reply patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch]: This is a patch, no touchy imports.

Comment (MercurialPony) [reply patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch]: So do I have to remove the unused imports or not?

File: src/main/java/net/minecraftforge/common/ForgeHooks.java
Comment (MercurialPony) [reply src/main/java/net/minecraftforge/common/ForgeHooks.java]: So I don't have to organize them?

File: patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch
Comment (Draco18s) [reply patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch]: In Minecraft vanilla files **do not ever touch imports, ever ever ever.** You have added several, REMOVE THEM.

File: src/main/java/net/minecraftforge/common/ForgeHooks.java
Comment (Draco18s) [reply src/main/java/net/minecraftforge/common/ForgeHooks.java]: No, don't organize them. (And you're right @covers1624 it's a Forge class, not a patch, I missed that, but it still needs to be fixed).

**Comment by MercurialPony** [comment]
@Draco18s I fixed everything, please approve. The question still stands: when will the additions be merged and released?

**Comment by bs2609** [comment]
You need to wait for mezz to review the PR. If they think it's ok, they'll assign it to Lex for final approval.

**Comment by Draco18s** [comment]
Looks much better now. Inclusion is primarily up to folks like Lex and mezz, who will review it at some point. Getting PRs added can be a lengthy process.

**Comment by MercurialPony** [comment]
Well thanks to everyone for the input. Here's hoping it will get added soon, I really need this for my mod.

**Comment by jabelar** [comment]
This is a useful event. It is a shame that more information about the context of why the knockback is occurring cannot be passed in. I understand that it is cleanest to hook into the knockback() method so you only have the parameters available that that method gets (plus the target entity in which the method is being called). It would be a lot cooler though if the DamageSource could be passed in as well, and also if you could determine if the knockback was due to blocking with shield. 

In other words, instead of firing the event inside the knockback() method it would be much more useful if it was fired outside the method and a bit more information passed along. 

But it is still pretty cool as is, and could always be extended in an unbreaking way later.

File: src/test/java/net/minecraftforge/debug/KnockBackHookTest.java
Comment (liach) [inline src/test/java/net/minecraftforge/debug/KnockBackHookTest.java]: Can you put 4 spaces for tabs?

File: src/main/java/net/minecraftforge/event/entity/living/LivingKnockBackEvent.java
Comment (liach) [inline src/main/java/net/minecraftforge/event/entity/living/LivingKnockBackEvent.java:58]:
```java
protected final Entity attacker;
protected float strength;
protected double ratioX, ratioZ;
protected final float originalStrength;
protected final double originalRatioX, originalRatioZ;

public LivingKnockBackEvent(EntityLivingBase target, Entity attacker, float strength, double ratioX, double ratioZ)
{
    super(target);
    this.attacker = attacker;
    this.originalStrength = strength;
    this.originalRatioX = ratioX;
    this.originalRatioZ = ratioZ;
    this.strength = strength;
    this.ratioX = ratioX;
    this.ratioZ = ratioZ;
}
```
In case of multiple mods modifying the same event.

File: src/test/java/net/minecraftforge/debug/KnockBackHookTest.java
Comment (mezz) [reply src/test/java/net/minecraftforge/debug/KnockBackHookTest.java]: As liach says, please use spaces for indentation.

File: src/main/java/net/minecraftforge/event/entity/living/LivingKnockBackEvent.java
Comment (mezz) [reply src/main/java/net/minecraftforge/event/entity/living/LivingKnockBackEvent.java:58]: This is a good suggestion from liach, please add the "original" values as final values, so mods using the events can have more information about modifications to the values.

**Review by mezz — CHANGES_REQUESTED** [review]
This is looking good. Please address the remaining comments and then I will test this out.
Consider jabelar's comment as well, and see if you can fire this in a location that gives it access to the damage source and if there is a shield. That will make the event useful in more situations.

**Comment by MercurialPony** [comment]
@mezz @jabelar  I'll get around to fixing the issues. However, I'm not sure how to go about passing more information to the event. I could add the event to every knockBack() method call, but that would require modifying at least 4 different classes and wouldn't work if other modders called the method. I could add more params to the knockBack() method, but I'm not sure that's a good idea for compatibility or similar issues.

**Comment by mezz** [comment]
I guess for now just keep it simple, the way it is.

**Comment by MercurialPony** [comment]
@mezz Alright, I've added everything you pointed out. Will this be added to the api now? Also, why do none of the other events with changeable params have original fields?

File: src/main/java/net/minecraftforge/event/entity/living/LivingKnockBackEvent.java
Comment (liach) [reply src/main/java/net/minecraftforge/event/entity/living/LivingKnockBackEvent.java:58]: Fixed now.

**Review by liach — APPROVED** [review]

**Review by mezz — APPROVED** [review]

**Comment by mezz** [comment]
Tested and works as expected.

**Pull request merged**
